File: python/4.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
import re


def part1(data):
    """ 2020 Day 4 Part 1

    >>> part1(['ecl:gry pid:860033327 eyr:2020 hcl:#fffffd', 'byr:1937 iyr:2017 cid:147 hgt:183cm', '', 'iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884', 'hcl:#cfa07d byr:1929', '', 'hcl:#ae17e1 iyr:2013', 'eyr:2024', 'ecl:brn pid:760753108 byr:1931', 'hgt:179cm', '', 'hcl:#cfa07d eyr:2025 pid:166559648', 'iyr:2011 ecl:brn hgt:59in', ''])
    2
    """

    passports = []
    currPassport = {}
    for line in data:
        if len(line) == 0:
            passports.append(currPassport)
            currPassport = {}
            continue

        for field in re.split(' ', line):
            k, v = re.split(':', field)
            currPassport[k] = v

    if currPassport:
        passports.append(currPassport)

    valid = 0
    for i, p in enumerate(passports):
        if len(p) == 8 or (len(p) == 7 and 'cid' not in p):
            valid += 1

    return valid


def part2(data):
    """ 2020 Day 4 Part 2

    >>> part2(['eyr:1972 cid:100', 'hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926', '', 'iyr:2019', 'hcl:#602927 eyr:1967 hgt:170cm', 'ecl:grn pid:012533040 byr:1946', '', 'hcl:dab227 iyr:2012', 'ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277', '', 'hgt:59cm ecl:zzz', 'eyr:2038 hcl:74454a iyr:2023', 'pid:3556412378 byr:2007', '', 'pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980', 'hcl:#623a2f', '', 'eyr:2029 ecl:blu cid:129 byr:1989', 'iyr:2014 pid:896056539 hcl:#a97842 hgt:165cm', '', 'hcl:#888785', 'hgt:164cm byr:2001 iyr:2015 cid:88', 'pid:545766238 ecl:hzl', 'eyr:2022', '', 'iyr:2010 hgt:158cm hcl:#b6652a ecl:blu byr:1944 eyr:2021 pid:093154719', ''])
    4
    """

    passports = []
    currPassport = {}
    for line in data:
        if len(line) == 0:
            passports.append(currPassport)
            currPassport = {}
            continue

        for field in re.split(' ', line):
            k, v = re.split(':', field)
            currPassport[k] = v

    if currPassport:
        passports.append(currPassport)

    valid = 0
    for p in passports:
        if len(p) == 8 or (len(p) == 7 and 'cid' not in p):
            try:
                if not (1920 <= int(p['byr']) <= 2002):
                    continue
                if not (2010 <= int(p['iyr']) <= 2020):
                    continue
                if not (2020 <= int(p['eyr']) <= 2030):
                    continue
                if ('cm' in p['hgt'] and not (150 <= int(p['hgt'][:-2]) <= 193)) or ('in' in p['hgt'] and not (59 <= int(p['hgt'][:-2]) <= 76)) or not ('cm' in p['hgt'] or 'in' in p['hgt']):
                    continue
                if not (p['hcl'][0] == '#' and int(p['hcl'][1:], 16) is not None):
                    continue
                if not (p['ecl'] in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']):
                    continue
                if not (len(p['pid']) == 9 and int(p['pid']) is not None):
                    continue

                valid += 1
            except ValueError:
                continue

    return valid


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nValid passports: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nValid passports with valid field data: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)

File: c/4.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#define defaultInput "../../Inputs/2020_4.txt"
#define dataLine 100


typedef struct passport_p1 {
	int birth, issue, expiration, height, pid, cid;
	char hair[20], eyes[20];
} Passport_p1;

typedef struct passport_p2 {
	int birth, issue, expiration, cid;
	char pid[20], height[20], hair[20], eyes[20];
} Passport_p2;

int findLines(char *fileName) {
	int numLines = 1;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            ++numLines;
        }
	}

	fclose(inFile);

	return numLines;
}

bool valid_p1(struct passport_p1 *passports, int i) {
    if (passports[i].birth == -1) {
        return false;
    } else if (passports[i].pid == -1) {
        return false;
    } else if (passports[i].issue == -1) {
        return false;
    } else if (passports[i].expiration == -1) {
        return false;
    } else if (passports[i].height == -1) {
        return false;
    } else if ((strcmp(passports[i].eyes, "\n")) == 0) {
        return false;
    } else if ((strcmp(passports[i].hair, "\n")) == 0) {
        return false;
    }

    return true;
}

bool valid_p2(struct passport_p2 *passports, int i) {
    if (passports[i].birth < 1920 || passports[i].birth > 2002) {
        return false;
    } else if (passports[i].issue < 2010 || passports[i].issue > 2020) {
        return false;
    } else if (passports[i].expiration < 2020 || passports[i].expiration > 2030) {
        return false;
    }

    int num = 0, j = 0;
    char temp[20];
    strcpy(temp, passports[i].height);
    char tempc = passports[i].height[j];

    while (isdigit(tempc) != 0) {
        num *= 10;
        num += tempc - 48;
        j++;
        tempc = passports[i].height[j];
    }

    if (tempc == 'c') {
        if (num < 150 || num > 193) {
            return false;
        }
    } else if (tempc == 'i') {
        if (num < 59 || num > 76) {
            return false;
        }
    } else {
        return false;
    }

    if (passports[i].hair[0] != '#') {
        return false;
    }

    int count = 0;
    j = 1;
    while (isxdigit(passports[i].hair[j]) != 0) {
        count++;
        j++;
    }

    if (count != 6) {
        return false;
    }

    if (!(strncmp(passports[i].eyes, "amb", 3) == 0 || strncmp(passports[i].eyes, "blu", 3) == 0 || strncmp(passports[i].eyes, "brn", 3) == 0 || strncmp(passports[i].eyes, "gry", 3) == 0 || strncmp(passports[i].eyes, "grn", 3) == 0 || strncmp(passports[i].eyes, "hzl", 3) == 0 || strncmp(passports[i].eyes, "oth", 3) == 0)) {
        return false;
    }

    count = 0;
    j = 0;
    while (isdigit(passports[i].pid[j]) != 0) {
        count++;
        j++;
    }

    if (count != 9) {
        return false;
    }

    return true;
}

int part1(int numLines, char *fileName) {
    int count = 0;

    struct passport_p1 *passports;

    passports = (Passport_p1 *)calloc(numLines, sizeof(Passport_p1));
    for (int i = 0; i < numLines; i++) {
        passports[i].birth = -1;
        passports[i].cid = -1;
        passports[i].expiration = -1;
        strcpy(passports[i].eyes, "\n");
        strcpy(passports[i].hair, "\n");
        passports[i].height = -1;
        passports[i].issue = -1;
        passports[i].pid = -1;
    }

    int i = 0;
	char textRead[dataLine], *p;

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            i++;
        }

        p = strtok(textRead, ":");

        while (p != NULL) {
            char field[5], data[20];
            strcpy(field, p);

            p = strtok(NULL, " ");
            if (p == NULL) {
                break;
            }
            strcpy(data, p);

            if (strncmp(field, "byr", 3) == 0) {
                passports[i].birth = atoi(data);
            } else if (strncmp(field, "iyr", 3) == 0) {
                passports[i].issue = atoi(data);
            } else if (strncmp(field, "eyr", 3) == 0) {
                passports[i].expiration = atoi(data);
            } else if (strncmp(field, "hgt", 3) == 0) {
                passports[i].height = atoi(data);
            } else if (strncmp(field, "hcl", 3) == 0) {
                strcpy(passports[i].hair, data);
            } else if (strncmp(field, "ecl", 3) == 0) {
                strcpy(passports[i].eyes, data);
            } else if (strncmp(field, "pid", 3) == 0) {
                passports[i].pid = atoi(data);
            } else if (strncmp(field, "cid", 3) == 0) {
                passports[i].cid = atoi(data);
            }

            p = strtok(NULL, ":");
        }
	}

	fclose(inFile);

    for (int i = 0; i < numLines; i++) {
        if (valid_p1(passports, i)) {
            count++;
        }
    }

    return count;
}

int part2(int numLines, char *fileName) {
    int count = 0;

    struct passport_p2 *passports;

    passports = (Passport_p2 *)calloc(numLines, sizeof(Passport_p2));
    for (int i = 0; i < numLines; i++) {
        passports[i].birth = -1;
        passports[i].cid = -1;
        passports[i].expiration = -1;
        strcpy(passports[i].eyes, "\n");
        strcpy(passports[i].hair, "\n");
        strcpy(passports[i].height, "\n");
        passports[i].issue = -1;
        strcpy(passports[i].pid, "\n");
    }

    int i = 0;
	char textRead[dataLine], *p;

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            i++;
        }

        p = strtok(textRead, ":");

        while (p != NULL) {
            char field[5], data[20];
            strcpy(field, p);

            p = strtok(NULL, " ");
            if (p == NULL) {
                break;
            }
            strcpy(data, p);

            if (strncmp(field, "byr", 3) == 0) {
                passports[i].birth = atoi(data);
            } else if (strncmp(field, "iyr", 3) == 0) {
                passports[i].issue = atoi(data);
            } else if (strncmp(field, "eyr", 3) == 0) {
                passports[i].expiration = atoi(data);
            } else if (strncmp(field, "hgt", 3) == 0) {
                strcpy(passports[i].height, data);
            } else if (strncmp(field, "hcl", 3) == 0) {
                strcpy(passports[i].hair, data);
            } else if (strncmp(field, "ecl", 3) == 0) {
                strcpy(passports[i].eyes, data);
            } else if (strncmp(field, "pid", 3) == 0) {
                strcpy(passports[i].pid, data);
            } else if (strncmp(field, "cid", 3) == 0) {
                passports[i].cid = atoi(data);
            }

            p = strtok(NULL, ":");
        }
	}

	fclose(inFile);

    for (int i = 0; i < numLines; i++) {
        if (valid_p2(passports, i)) {
            count++;
        }
    }

    return count;
}

int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

	int numLines, count = 0;
	struct passport *passports;

	numLines = findLines(inputPath);

    if (numLines == -1) {
		printf("Error: could not open file. Quitting\n");
		return -1;
	}

    clock_t t; 
    t = clock(); 
    int p1 = part1(numLines, inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nValid passports: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock();
    int p2 = part2(numLines, inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nValid passports: %d\nRan in %f seconds\n", p2, t_p2);

	return 0;
}
